sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,814 @@
+  implementation defined non-zero value.
+* A memory pool has a *threshold size* (in bytes) which dictates an amount of
+  the user would like the SYCL runtime to maintain in the pool, even when not
+  allocated out to a USM pointer. The threshold size is a hint, and so an
+  implementation should try to maintain memory in the pool to this value, but it
+  is not required to do so. The default threshold size is `0`.
+* A memory pool has a *current size* (in bytes) which tracks the total memory
+  which is allocated to the pool.
+* A memory pool has a *free size* (in bytes) which tracks the amount of the
+  free memory; total memory which has not been allocated to a USM pointer, and
+  is still available to be allocated.
+* A memory pool has a *allocation kind* which is a value of `usm::alloc` which
+  dictates the type of USM memory that is allocated to the pool.
+* A memory pool is associated with a context and one or more device(s),
+  depending on the allocation kind.
+
+A memory pool and the asynchronous malloc and free commands which interact with
+them adhere to the following behaviors:
+
+* When an asynchronous malloc command is enqueued a valid USM pointer of the
+  allocation kind is returned immediately. The asynchronous malloc command may
+  execute any time between the command being enqueued and the command being
+  synchronized with. The pointer returned is not valid until the asynchronous
+  malloc command has completed.
+* When an asynchronous malloc command executes it will allocate USM memory of
+  the allocation kind from the memory pool.
+* When a USM pointer is being allocated from the memory pool, it may allocate
+  from free memory in the pool, including memory which was previously allocated
+  to a USM pointer and has been freed or it may opportunistically re-use memory
+  allocated to an existing USM pointer, for which an asynchronous free command
+  has been enqueued to free, and that the asynchronous malloc command is
+  dependent on.
+* Additional memory can be allocated to the memory pool when required for an
+  asynchronous malloc command or otherwise. Allocations to the pool must be in a
+  multiple of the allocation chunk size.
+* When an asynchronous free command is enqueued, the USM pointer to be freed is
+  immediately no longer valid. The asynchronous free command may execute any
+  time between the command being enqueued and the command being synchronized
+  with. Once the asynchronous free command is enqueued memory allocated to that
+  USM pointer is now free for the memory pool to re-allocate to another USM
+  pointer either once the command has completed, or if an asynchronous command
+  is dependent on this asynchronous free command.
+* The memory pool will not release memory in the pool that is currently
+  allocated to a USM pointer, however, it may release any free memory in the
+  pool. When the memory pool releases memory from the pool, this is done so in
+  multiples of the allocation chunk size.
```

Comment:
This seems like an implementation detail too.

Reply:
I agree, I've moved this to the implementation notes sections in https://github.com/intel/llvm/pull/14800/commits/2898549e9ad8c0f9c947536cf7f993cb44aa928b.